infra/Pos/Reporting/Methods.hs

```diff
@@ -61,6 +72,125 @@ import           Pos.Reporting.MemState (HasLoggerConfig (..), HasReportServers
+-- error, because there's no possibility to have two handles on the
+-- same file, see 'System.IO' documentation on handles. See
+-- 'withTempLogFile' to overcome this problem.
+sendReport
+    :: (HasConfiguration, HasCompileInfo, MonadIO m, MonadMask m)
+    => [FilePath]                 -- ^ Log files to read from
+    -> ReportType
+    -> Text
+    -> String
+    -> m ()
+sendReport logFiles reportType appName reportServerUri = do
+    curTime <- liftIO getCurrentTime
+    existingFiles <- filterM (liftIO . doesFileExist) logFiles
+    when (null existingFiles && not (null logFiles)) $
+        throwM $ CantRetrieveLogs logFiles
+
+    rq0 <- parseUrlThrow $ reportServerUri <//> "report"
+    let pathsPart = map partFile' existingFiles
+    let payloadPart =
+            Form.partLBS "payload"
+            (encode $ mkReportInfo curTime)
+    -- If performance will ever be a concern, moving to a global manager
+    -- should help a lot.
+    reportManager <- liftIO $ newManager tlsManagerSettings
+
+    -- Assemble the `Request` out of the Form data.
+    rq <- Form.formDataBody (payloadPart : pathsPart) rq0
+
+    -- Actually perform the HTTP `Request`.
+    e  <- try $ liftIO $ httpLbs rq reportManager
+    whenLeft e $ \(e' :: SomeException) -> throwM $ SendingError e'
+  where
+    partFile' fp = Form.partFile (toFileName fp) fp
+    toFileName = toText . takeFileName
+    mkReportInfo curTime =
+        ReportInfo
+        { rApplication = appName
+        -- We are using version of 'cardano-sl-infra' here. We agreed
+        -- that the version of 'cardano-sl' and it subpackages should
+        -- be same.
+        , rVersion = version
+        , rBuild = pretty compileInfo
+        , rOS = toText (os <> "-" <> arch)
+        , rMagic = getProtocolMagic protocolMagic
+        , rDate = curTime
+        , rReportType = reportType
+        }
+
+
+-- | Given logger config, retrieves all (logger name, filepath) for
+-- every logger that has file handle. Filepath inside does __not__
+-- contain the common logger config prefix.
+retrieveLogFiles :: LoggerConfig -> [([Text], FilePath)]
+retrieveLogFiles lconfig = fromLogTree $ lconfig ^. lcTree
+  where
+    fromLogTree lt =
+        let curElems = map ([],) (lt ^.. ltFiles . each . hwFilePath)
+            addFoo (part, node) = map (first (part :)) $ fromLogTree node
+        in curElems ++ concatMap addFoo (lt ^. ltSubloggers . to HM.toList)
+
+-- | Pass a list of absolute paths to log files. This function will
+-- archive and compress these files and put resulting file into log
+-- directory (returning filepath is absolute).
+compressLogs :: (MonadIO m) => [FilePath] -> m FilePath
+compressLogs files = liftIO $ do
+    tar <- tarPackIndependently files
+    tarxz <-
+        BS.concat <$>
+        runConduitRes (yield tar .| Lzma.compress (Just 0) .| consume)
+    aName <- getArchiveName
+    withFile aName WriteMode $ \handle -> do
+        BS.hPut handle tarxz
+        hFlush handle
```

Comment:
won't `withFile` close and flush the file along with closing?

Reply:
`hClose` should do `hFlush` inside, but I'm not sure some kind of weird laziness won't kick in, so let's leave it as it is.